[WIP] JSON Deserializer

matthiasbeyer opened the pull request:
Targets #151

This is highly WIP. Please review the interface definitions before I
start implementing this stuff!

matthiasbeyer committed:
Add common header for yajl backend

matthiasbeyer committed:
Add common.c for yajl JSON backend including stubs

matthiasbeyer committed:
Add allocation function callbacks including stubs

matthiasbeyer committed:
Add allocation func table

matthiasbeyer committed:
Add implementation for serializer_yajl_state_new()

matthiasbeyer committed:
Add implementation of serializer_yajl_state_free()

matthiasbeyer committed:
Add JSON deserializer backend interface definition

matthiasbeyer committed:
Add deserializer file with stubs

matthiasbeyer committed:
Add yajl callbacks definitions and stubs

matthiasbeyer committed:
Add YAJL callback table

matthiasbeyer committed:
Add implementation for ws_serializer_json_deserializer_new()

matthiasbeyer committed:
Add deserialize callback with stub

neithernut commented on src/serialize/json/common.c:
```diff
@@ -0,0 +1,140 @@
+ *
+ */
+
+void *
```
static keyword missing, also for all the other funcs.

neithernut commented on src/serialize/json/common.c:
```diff
@@ -0,0 +1,140 @@
+    state->handle = yajl_alloc(callbacks, &YAJL_ALLOC_FUNCS, NULL);
+
+    if (!yajl_config(state->handle, YAJL_OPTS)) {
+        //!< @todo yajl config failed. What to do ?
```
free and return an error

neithernut commented on src/serialize/json/deserializer.c:
```diff
@@ -0,0 +1,330 @@
+ *
+ * @return zero on failure, else nonzero value
+ */
+int
```
again `static` missing.

neithernut commented on src/serialize/json/deserializer.c:
```diff
@@ -0,0 +1,330 @@
+ *
+ */
+
+yajl_callbacks YAJL_CALLBACKS = {
```
Should also be `static`.

neithernut commented on src/serialize/json/deserializer.c:
```diff
@@ -0,0 +1,330 @@
+ *
+ * @return negative error value on error, else number of deserialized bytes
+ */
+ssize_t
```
`static` missing here, too.

matthiasbeyer committed:
Pass deserializer as context around

matthiasbeyer committed:
Pass size to yajl state object allocation

This way, the yajl state object becomes an abstract one

matthiasbeyer committed:
Add deserializer state type including allocator helper

matthiasbeyer committed:
Add start_map callback implementation

matthiasbeyer committed:
Add (partial) implementation of string callback

neithernut commented on src/serialize/json/deserializer.c:
```diff
@@ -0,0 +1,433 @@
+    struct {
+        const unsigned char* uid;
+        size_t len;
+        bool set;
```
set `uid` to `NULL` initially and you don't need this flag.

neithernut commented on src/serialize/json/deserializer.c:
```diff
@@ -0,0 +1,433 @@
+    switch (state->current_state) {
+    case STATE_INIT:
+        // a strange thing happened
+        return 0;
```
You cluster all of these and return something _other_ than `0` to **halt** the parser.

matthiasbeyer replied:
From the yajl documentation:

>  If zero, the parse will be canceled

neithernut commented on src/serialize/json/deserializer.c:
```diff
@@ -0,0 +1,433 @@
+    case STATE_UID:
+        state->uid.uid = str;
+        state->uid.len = len;
+        state->uid.set = true;
```
See note above.

matthiasbeyer committed:
Outsource states into own file

matthiasbeyer committed:
Replace strcmp() "switchcase" with jump helper function in string cb

matthiasbeyer committed:
Add documentation about deserializer and statemachine

matthiasbeyer committed:
Moved callbacks to own file

matthiasbeyer committed:
Outsource deserializer state

matthiasbeyer committed:
Rename: d_yajl_state -> deserializer_state

matthiasbeyer committed:
Add documentation to deserializer state object type

matthiasbeyer committed:
Fix: Go back to message state after processing UID/Type string

matthiasbeyer committed:
Add command buffer in internal state object type

matthiasbeyer committed:
Add deserialize_state_init_cmds()

matthiasbeyer committed:
Add command push method for state object type

matthiasbeyer committed:
Add implementation for start_array callback

matthiasbeyer committed:
Add case for running into a new command

matthiasbeyer committed:
Add case for running into the key of the next command

matthiasbeyer committed:
Add state net for init->command arguments

matthiasbeyer committed:
Add command argument array parse start entry

matthiasbeyer committed:
Add command argument parsing state net

matthiasbeyer committed:
Fix: ws_command name member should be setable

matthiasbeyer commented on src/command/command.h:
```diff
@@ -86,7 +86,7 @@ typedef int (*ws_special_command_func)(struct ws_command_processor*,
  */
 struct ws_command {
-    char const* const name; //!< @public name of the command
+    char* name; //!< @public name of the command
```
@neithernut I had to modify this to be able to set the name of the command without doing ugly casts. Is this appropriate? SYN.

neithernut replied:
Nope, `char const*` is ok, but plain `char*` is not.
Do a cast instead.

matthiasbeyer committed:
Add helpers for processing arguments + stubs

matthiasbeyer committed:
Add implementation for deserialize_state_new_arg()

matthiasbeyer committed:
Add implementation for deserialize_state_mod_last_arg_set_direct()

matthiasbeyer committed:
Add implementation for deserialize_state_mod_last_arg_set_indirect()

matthiasbeyer committed:
Add: Command array initialization

matthiasbeyer committed:
Add command arg array parsing close state transition, including doc

matthiasbeyer committed:
Add transition: cmd parsing -> cmd array state

From the command parsing state to the command array state, including the
finalizing of the command.

Documentation included.

matthiasbeyer commented:
I'm ready for today, can I get a careful (and calmly) review from you, @neithernut and maybe even @TheNeikos ?

What I'm currently not focusing on: Cleaning up if something wents wrong. I solely focus on the plain parsing. I also do not focus on speed, there are some `realloc(size+1)` calls in the code etc etc, as I try to keep the things as simple as possible by now!

matthiasbeyer committed:
Added enum member documentation

matthiasbeyer committed:
Add transition doc: Command Arguments -> Commands

matthiasbeyer committed:
Added Key parsing callback implementation

This callback handles the keys of the JSON object, the code was
previousely located in the string parser, that was wrong.

The code was also fixed by this move (variable rename from `str`
to `key`).

matthiasbeyer committed:
Do not allocate a new arg when starting parsing args

Each parsing step should allocate an own argument, it's less overhead
this way, as the final argument parsing step (the "hey, I'm ready with
the args" one) does not have to deallocate the last argument object.

matthiasbeyer committed:
Fix: direct/indirect was interchanged

matthiasbeyer committed:
Fix: We must modify the _last_, not the one before

matthiasbeyer committed:
Added implementation for indirect arguments

This adds the implementation of the yajl_number_cb() function callback
with code for a indirect value in the command argument array.

matthiasbeyer committed:
Add code for starting a new direct argument

matthiasbeyer committed:
Add code for parsing the map key for direct argument values

This patch includes the new states.

matthiasbeyer committed:
Add documentation of direct/indirect arg state jumping

matthiasbeyer committed:
Fix: yajl option setting

matthiasbeyer committed:
Add indirect command argument parsing state net with expl.

matthiasbeyer committed:
Add note on argument parsing

neithernut commented:
I'm finally in front of a terminal again and can now start reviewing this.

neithernut commented on doc/custom/serializer/json/statemachine.md:
```diff
@@ -0,0 +1,123 @@
+The statemachine in the JSON deserializer {#json_statemachine}
+=========================================
+
+This chapter documents the statemachine which is used in the JSON deserializer
```
It's "state machine".

matthiasbeyer committed:
Add transition from command array parsing back to message state

neithernut commented on doc/custom/serializer/json/statemachine.md:
```diff
@@ -0,0 +1,123 @@
+State diagrams
+--------------
+
+This section contains state diagrams for single paths through the overall state
```
I don't see any purpose in this section.

neithernut commented on doc/custom/serializer/json/statemachine.md:
```diff
@@ -0,0 +1,123 @@
+                |              | "["
+                |              |
+                |              v            "{"   +--------------------------+
+                \------ Command Arguments ------> | Command argument parsing |
```
The transition back to the initial state is missing.

neithernut commented on src/serialize/json/common.c:
```diff
@@ -0,0 +1,148 @@
+yajl_alloc_funcs YAJL_ALLOC_FUNCS = {
+    .malloc     = yajl_malloc_func_cb,
+    .realloc    = yajl_realloc_func_cb,
+    .free       = yajl_free_func_cb,
```
_please_ check when and how these functions get involved and whether you can pass `NULL`. Having a `realloc()` function returning just `NULL` is completely nuts!

matthiasbeyer replied:
Indeed. I will check.

neithernut commented on src/serialize/json/common.c:
```diff
@@ -0,0 +1,148 @@
+            yajl_free(state->handle);
+            return NULL;
+        }
+    }
```
Is there any point in _not_ just OR'ing the `yajl_config()` involcations but have the attributes stored in an array?

matthiasbeyer replied:
Yes. From the example on the yajl homepage, it seems that you _have to_ pass each option on its own and say "I want to set this" with passing the `1` in the arguments. Completely stupid, if you ask me, but the example does it this way (also because in the example, you can set the options via cmdline).

Things can be optimized later, first this stuff should work.

neithernut replied:
I think you misunderstood. I meant:

``` C
if (!(yajl_config(state, foo, 1) &&
      yajl_config(state, bar, 1) &&
...
) {
    //failure.
}
```

neithernut commented on src/serialize/json/deserializer_callbacks.c:
```diff
@@ -0,0 +1,344 @@
+    switch (state->current_state) {
+    case STATE_INIT:
+        // a strange thing happened
+        return 0;
```
This is caught in `default` already.

neithernut commented on src/serialize/json/deserializer_callbacks.c:
```diff
@@ -0,0 +1,344 @@
+        {
+            enum json_backend_state next;
+            next = get_next_state_for_string(STATE_MSG, key);
+            if (next == STATE_INIT) {
```
`STATE_INIT`? Please introduce `STATE_INVALID` or something.

neithernut commented on src/serialize/json/deserializer_callbacks.c:
```diff
@@ -0,0 +1,344 @@
+                    state->current_state = next_state[t];
+                }
+                break;
+            }
```
A switch-case, embedded in another switch-case? And even a scope in it?

matthiasbeyer replied:
Will be "optimized" away as soon as things work.

neithernut commented on src/serialize/json/deserializer_callbacks.c:
```diff
@@ -0,0 +1,344 @@
+    switch (state->current_state) {
+    case STATE_COMMAND_ARY_NEW_COMMAND:
+        // We are ready with the command parsing for one command now. Lets
+        // finalizy the temporary stuff and go back to the command array state.
```
Let's finally what? The temporary stuff?
Plus points for accidential meme reference, but I _do_ think you want to fix this comment.

neithernut commented on src/serialize/json/deserializer_callbacks.c:
```diff
@@ -0,0 +1,344 @@
+    default:
+        break;
+
+    }
```
Also, you forgot the `end_map` of the message itself.

matthiasbeyer replied:
What do you mean exactly? Note: I'm not completely done and this is WIP, so things are missing, of course.

neithernut replied:
The number of cases in a `switch` case are directly related to the number of array types there are. If they are not equal fro `begin_array` and `end_array`, wou certainly fucked things up.

matthiasbeyer committed:
Add extension to Argument parsing state machine net

neithernut commented on src/serialize/json/deserializer_callbacks.c:
```diff
@@ -0,0 +1,344 @@
+        break;
+
+    default:
+        break;
```
I think you meant to write `return 0;` here.

neithernut commented on src/serialize/json/deserializer_callbacks.c:
```diff
@@ -0,0 +1,344 @@
+        break;
+
+    default:
+        break;
```
Same as above.

neithernut commented on src/serialize/json/deserializer_callbacks.c:
```diff
@@ -0,0 +1,344 @@
+    default:
+        break;
+
+    }
```
You have two distinct ways for entering an array, but only one way out?

neithernut commented on src/serialize/json/deserializer_state.c:
```diff
@@ -0,0 +1,143 @@
+
+        state->tmp_command_args.vals = NULL;
+        state->tmp_command_args.num = 0;
+    }
```
**Early. Return.** vs. _the giant intended list_ and shenigans.

neithernut commented on src/serialize/json/deserializer_state.h:
```diff
@@ -0,0 +1,172 @@
+        struct ws_statement** cmdbuf;
+        size_t buflen;
+        size_t next;
+    } commands;
```
You know you yould also just use a `wl_array`?

matthiasbeyer replied:
`wl` or `ws`? (I know, both are possible, but which one would you prefer?)

neithernut replied:
`wl`!

matthiasbeyer committed:
Fix: use early return in deserialize_state_new()

Because neithernut loses his poise on this.

neithernut commented on src/serialize/json/deserializer_state.c:
```diff
@@ -0,0 +1,143 @@
+    memset(&s->tmp_command_args, 0, sizeof(s->tmp_command_args));
+
+    return 0;
+}
```
You wouldn't have to do this if you used `wl_array`.

neithernut commented on src/values/value_type.h:
```diff
@@ -49,6 +49,18 @@ enum ws_value_type {
+enum ws_value_type
+ws_value_type_from_value_name(
+    char const* name
+);
```
Could you push this as a seperate PR?

matthiasbeyer replied:
Will do asap.

neithernut commented on doc/custom/serializer/json/statemachine.md:
```diff
@@ -0,0 +1,124 @@
+             \----------/
+
+(*) The used strings are as constants in the code and are defined in a central
+place.
```
I would put this entire file into the appropriate source directory. This way, you see the documentation directly when browsing sources. You can also view the documentation live on github without clicking trough half of the source tree.

matthiasbeyer replied:
I wouldn't. That's what the `doc/custom` directory is for and why doxygen has a `@see` tag.

neithernut replied:
It makes the life of someone looking for documentation a _whole lot_ easier, especially if they do not know much about doxygen.

matthiasbeyer replied:
And it bothers people who know about how things work. Documentation belongs to source code, but now _that_ close.

neithernut replied:
Yes, it does.

matthiasbeyer committed:
Add call state transition for {"nil":null}

matthiasbeyer committed:
Add helper for making last argument a NIL one

matthiasbeyer committed:
Add helper for setting last arg to bool

matthiasbeyer committed:
Add bool argument transition

matthiasbeyer committed:
Add indirect argument (int) argument passing

matthiasbeyer committed:
Add helper for set last argument to int

matthiasbeyer committed:
Add transition for parsing direct int

matthiasbeyer committed:
Add transition for integer indirect args per yajl number callback

matthiasbeyer committed:
Add helper for setting last argument to string

matthiasbeyer committed:
Add transition for string arguments

matthiasbeyer committed:
Add missing break in new arg case

matthiasbeyer committed:
Fix return values

matthiasbeyer committed:
Move documentation into src/

matthiasbeyer commented:
Short summary what is missing by now:
- Deserialize starting callback (The deserializer does nothing by now, as the deserializing is never started)
- Objects as arguments (waiting on `ws_value_obj_id` or however the typename is)
- Arrays as arguments (How to implement this, anyways? Sets?)
- Sets as arguments
- Pretty much everything besides `null`, `true`, `false`, the appropriate JSON-objects (`{"nil":null}` and so on), integers and integer-values as arguments 
- Propably a lot more.

Feel free to add more points.

matthiasbeyer committed:
Add implementation for deserialize()

matthiasbeyer committed:
Transform internal deserializer state processing helpers to wl_array

matthiasbeyer commented:
I will close this now unmerged, as this is a load of crap and other PRs for more clean history etc are in progress.
